_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+        userResponse = {
+            ...defaultUserSchema,
+            id: `${user.id}`,
+            userName: user.email,
+            name: {
+                givenName,
+                familyName
+            },
+            emails: [{
+                primary: true,
+                value: user.email,
+                type: "work"
+            }],
+            displayName: name,
+            externalId: user.externalId,
+            active: user.active
+        };
+    }
+    res.status(httpStatus).json(userResponse);
+});
+```
+
+With that code saved in `scim.ts`, we have the first endpoint we can interact with. Now is the best time to test our code to ensure we are on the right track. 
+
+To test our code, we'll be using [Postman](https://www.postman.com/) to make requests to our SCIM server. I'll also be adding an external tool, [Morgan](https://github.com/expressjs/morgan), that will help log HTTP requests; by doing so, we'll be able to confirm whether requests are reaching our server. 
+
+### Test with Postman
+
+We'll need to set up a few things for Postman to authenticate and interact with our local server. 
+
+#### 1 ) Add Bearer Token Auth to Secure the SCIM Routes
+
+When testing our routes, Postman will authenticate to the Todo app backend using a bearer token. To support token auth, first install the appropriate passport libraries. 
+
+In your terminal, run: 
+
+```
+npm install passport-http-bearer
+npm install @types/passport-http-bearer -D
+```
+
+Then import `passportBearer` and create a token auth strategy in `main.ts`: 
+
+```
+import passportBearer from 'passport-http-bearer';
+const BearerStrategy = passportBearer.Strategy;
+
+passport.use(new BearerStrategy(
+  async (apikey, done) => {
+    const org = await prisma.org.findFirst({
+      where: {
+        apikey: apikey
+      }
+    });
+
+    return done(null, org);
+  }
+));
+```
+
+In `scim.ts`, import Passport:
+
+```
+import passport from 'passport';
+```
+
+And add passport's bearer token auth to each SCIM route in `scim.ts`. For example,
+
+```
+scimRoute.post('/Users', passport.authenticate('bearer'), async (req, res)
+```
+
+#### 2) Support Okta's SCIM content-type headers
+
+You'll also need to install [body-parser](https://www.npmjs.com/package/body-parser) for Express.js to accept `content-type application/scim+json`, because Okta sends this in the request headers instead of `content-type application/json`. We'll need this to read the request body from Okta. Note: This specific content-type header is required by [the SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.1)
+
+On the command line, `npm install passport-http-bearer` to add the library to your project.
+
+In `scim.ts`, import `body-parser`:
+```
+import bodyParser from 'body-parser';
+```
+And instruct the app to use `bodyParser` for the relevant content types: 
+
+```
+app.use(bodyParser.json({ type: ['application/json', 'application/scim+json'] }));
+```
```

Comment:
@edunham can you please edit this to app.use(bodyParser.json({ type: 'application/scim+json' })); - basically no need to mention 'application/json'